[Structural] Use Gauss rule in solid shell element

rubenzorrilla opened the pull request:
**📝 Description**
In line to what has been discussed in https://github.com/KratosMultiphysics/Kratos/issues/13085, this PR targets arranging the KratosCore integration rules. Specifically, this PR removes the unique use of the problematic extended Gauss rules, that is from the solid shell element.

@loumalouomega I substituted the extended Gauss by the standard ones. Once we have the Lobatto ones properly implemented, we can update it to these if you consider it more appropriate. 

PD: el put* elemento maligno ataca de nuevo.

rubenzorrilla committed:
Use Gauss rule in solid shell element

loumalouomega reviewed (changes requested):
You cannot simply use standard gauss points. The model has specific integration and can be manually replaced, I will do, but please don't use the standard ones

loumalouomega committed:
Merge branch 'master' into structural/gauss-rule-in-solid-shell

loumalouomega committed:
Refactor SolidShellElementSprism3D6N: Replace typedefs with using declarations and enhance documentation

- Replaced typedefs with using declarations for improved readability and modern C++ style.
- Enhanced documentation for constructors, methods, and member variables to clarify functionality and usage.
- Corrected typos in comments and improved descriptions for enums and methods.
- Added static methods for integration points and shape functions to streamline calculations.

loumalouomega committed:
Merge branch 'master' into structural/gauss-rule-in-solid-shell

loumalouomega committed:
Fix

matekelemen reviewed (changes requested):
@loumalouomega you shouldn't approve a PR you contributed to.

Also, please do an actual review on something you approve.

rubenzorrilla commented on applications/StructuralMechanicsApplication/custom_elements/solid_elements/solid_shell_element_sprism_3D6N.h:
```diff
@@ -1423,27 +1448,89 @@ class KRATOS_API(STRUCTURAL_MECHANICS_APPLICATION) SolidShellElementSprism3D6N
+     * integration point corresponding to the different Gauss integration methods.
+     */
+    static const ShapeFunctionsLocalGradientsContainerType AllShapeFunctionsLocalGradients();
+
```
I'd keep these `private` to avoid possible misuses.

loumalouomega replied:
Okay

rubenzorrilla commented on applications/StructuralMechanicsApplication/custom_elements/solid_elements/solid_shell_element_sprism_3D6N.cpp:
```diff
@@ -4011,26 +4145,159 @@ void SolidShellElementSprism3D6N::CalculateVolumeForce(
+            CalculateShapeFunctionsIntegrationPointsValues(GeometryData::IntegrationMethod::GI_GAUSS_2 ),
+            CalculateShapeFunctionsIntegrationPointsValues(GeometryData::IntegrationMethod::GI_GAUSS_3 ),
+            CalculateShapeFunctionsIntegrationPointsValues(GeometryData::IntegrationMethod::GI_GAUSS_4 ),
+            CalculateShapeFunctionsIntegrationPointsValues(GeometryData::IntegrationMethod::GI_GAUSS_5 )
```
If you do this we end up with the same mess but in another place... `GI_GAUSS_X` does not refer to Gauss-Legendre integration (as expected) but to a different thing.

I'd just simply hardcode the integration rule you actually need, and remove the possibility to change it as this, as far as I understand, shouldn't be required by the user. Am I wrong?

loumalouomega replied:
The integration is hardcoded, can be even more hardcoded, but no one will see it. It is not general for all geometries. GeometryData::IntegrationMethod::GI_GAUSS_1 is just an enu,I can simply use a integer.

loumalouomega replied:
Here you are, now is fully hardcoded

rubenzorrilla replied:
Exactly. The deeper the better in this case.

rubenzorrilla replied:
Can we please avoid the use of `GeometryData::IntegrationMethod`? The point is that this might be misleading when browsing the code in the future (i.e., I'd expect the standard Gauss rule from the core but we are using a "custom" one in here).

loumalouomega replied:
I can use an int. It is an enum

![1840c968ae60de0911248c5e334e8ca8de67eaa752c37523812d84409a4e0ec3_1](https://github.com/user-attachments/assets/0262d83e-2c5e-422d-af2f-4f5f82874c74)

rubenzorrilla replied:
Yes. So instead of `GI_GAUSS_1` just put `1`. Though equivalent, the message one gets when going through or browsing the code is different.

loumalouomega replied:
It must start at 0. It is not stored as a map, but as an array, so must be ordered and starting at 0.

rubenzorrilla replied:
Sure.

loumalouomega committed:
Fix variable types in SolidShellElementSprism3D6N for consistency and clarity

loumalouomega committed:
Make it private

loumalouomega committed:
More hardcoded

loumalouomega committed:
Remove extended prism quadratures

rubenzorrilla commented on applications/StructuralMechanicsApplication/custom_elements/solid_elements/solid_shell_element_sprism_3D6N.cpp:
```diff
@@ -4011,26 +4142,159 @@ void SolidShellElementSprism3D6N::CalculateVolumeForce(
+            CalculateShapeFunctionsIntegrationPointsValues(GeometryData::IntegrationMethod::GI_GAUSS_2 ),
+            CalculateShapeFunctionsIntegrationPointsValues(GeometryData::IntegrationMethod::GI_GAUSS_3 ),
+            CalculateShapeFunctionsIntegrationPointsValues(GeometryData::IntegrationMethod::GI_GAUSS_4 ),
+            CalculateShapeFunctionsIntegrationPointsValues(GeometryData::IntegrationMethod::GI_GAUSS_5 )
```
```suggestion
            CalculateShapeFunctionsIntegrationPointsValues(1),
            CalculateShapeFunctionsIntegrationPointsValues(2),
            CalculateShapeFunctionsIntegrationPointsValues(3),
            CalculateShapeFunctionsIntegrationPointsValues(4),
            CalculateShapeFunctionsIntegrationPointsValues(5)
```
Like this.

loumalouomega replied:
As I said, must start at 0.

loumalouomega committed:
Refactor SolidShellElementSprism3D6N: Improve code readability and consistency

- Removed unused IntegrationMethod type alias.
- Reformatted constructor signatures for better readability.
- Updated parameter names for consistency (e.g., `ThisNodes` to `rNodes`).
- Added a new method to return the integration method used.
- Improved documentation comments for clarity.
- Corrected typos in comments and variable names.
- Enhanced code structure by grouping related methods and variables.

loumalouomega committed:
Add GetDeltaDisplacement method to calculate delta displacements

matekelemen reviewed (approved):
Fine.

I'll just stress my dislike for, and the dangers of, self-approved changes ending up so easily in master.
